crates/astria-composer/src/searcher/executor.rs

```diff
@@ -120,178 +139,266 @@ impl Executor {
+                }
+                .instrument(span)
+            },
         );
+    tryhard::retry_fn(|| {
+        let client = client.clone();
+        let tx = tx.clone();
+        let span = info_span!(parent: span.clone(), "attempt submission");
+        async move { client.submit_transaction_sync(tx).await }.instrument(span)
+    })
+    .with_config(retry_config)
+    .await
+    .wrap_err("failed submitting a signed transaction after 1024 attempts")
+}
 
-        self.status.send_modify(|status| {
-            status.is_connected = true;
-        });
-        Ok(())
+// TODO: zeroize this
+pin_project! {
+    /// A future to submit a bundle to the sequencer, returning the next nonce that should be used for the next submission.
+    ///
+    /// The future will fetch a new nonce from the sequencer if a submission returned an `INVALID_NONCE` error code.
+    ///
+    /// The future will only return an error if it ultimately failed submitting a transaction due to the underlying
+    /// transport failing. This can be taken as a break condition to exit the executor loop.
+    ///
+    /// If the sequencer returned a non-zero abci code (albeit not `INVALID_NONCE`), this future will return with
+    /// that nonce it used to submit the non-zero abci code request.
+    struct SubmitFut {
+        client: sequencer_client::HttpClient,
+        address: Address,
+        nonce: u32,
+        signing_key: SigningKey,
+        #[pin]
+        state: SubmitState,
+        bundle: Vec<Action>,
     }
 }
-/// A thin wrapper around [`sequencer_client::Client`] to add timeouts.
-///
-/// Currently only provides a timeout for `abci_info`.
-#[derive(Clone, Debug)]
-struct SequencerClient {
-    inner: sequencer_client::HttpClient,
-}
 
-impl SequencerClient {
-    #[instrument]
-    fn new(url: &str) -> eyre::Result<Self> {
-        let inner = sequencer_client::HttpClient::new(url)
-            .wrap_err("failed to construct sequencer client")?;
-        Ok(Self {
-            inner,
-        })
+pin_project! {
+    #[project = SubmitStateProj]
+    enum SubmitState {
+        NotStarted,
+        WaitingForSubmit {
+            #[pin]
+            fut: Pin<Box<dyn Future<Output = eyre::Result<tx_sync::Response>> + Send>>,
+        },
+        WaitingForNonce {
+            #[pin]
+            fut: Pin<Box<dyn Future<Output = eyre::Result<u32>> + Send>>,
+        }
     }
+}
 
-    /// Wrapper around [`Client::get_latest_nonce`] with a 1s timeout.
-    async fn get_latest_nonce(&self, address: Address) -> eyre::Result<NonceResponse> {
-        tokio::time::timeout(
-            Duration::from_secs(1),
-            self.inner.get_latest_nonce(address.0),
-        )
-        .await
-        .wrap_err("request timed out")?
-        .wrap_err("RPC returned with error")
-    }
+impl Future for SubmitFut {
+    type Output = eyre::Result<u32>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        loop {
```

Comment:
is it ok to loop in a poll function like this? doesn't this block the async runtime? i thought you're supposed to return `Poll::Pending` in the cases where the future isn't ready (for example here, if the `waiting_for_nonce` future is pending)

Reply:
i guess this *should* be fine because eventually the tx will submit and we'll get a response which will return `Poll::Ready`, but i'm suspect of the case where we end up stuck in a loop of continuously submitting with invalid nonces.

Reply:
The bit that makes this all work is this" https://docs.rs/futures/latest/futures/macro.ready.html

Have a look at the source: the loop returns early if the currently polled futures aren't ready

Reply:
> i guess this *should* be fine because eventually the tx will submit and we'll get a response which will return `Poll::Ready`, but i'm suspect of the case where we end up stuck in a loop of continuously submitting with invalid nonces.

Just saw the edit after you submitted:

Yeah, that would indeed happen. However, that case would be in principle fine in a service. It would not do a lot of useful work but continuously try and resubmit.

The point here is that it shouldn't be too smart: if it gets a nonce from sequencer, then tries to submit the tx, but gets told the nonce is invalid, then something is wrong in other services, but it's ok from the pov of composer